### PR TITLE
Optimizing Blueprints Loading 

### PR DESCRIPTION
**Pull request by HUSSAR-mtrela**
- [x] optimize table.hash function (+30% performance)
- [x] optimize bluprints.lua to save bp.CategoriesHash instead of discarding it (+32% performance)
- [x] optimize sim lua files to use bp.CategoriesHash instead of `for loops` to look up bp.Categories
- [x] optimize units analyzer to use bp.CategoriesHash instead of looking up bp.Categories
- [ ] optimize lobby.lua to pre-load bps in smaller chunks instead of one big task
  
  **Note this PR is in progress and it will be ready when all above task are complete**

### CODE REVIEW

**Comment by aeoncleanse** [comment]
@ChrisKitching - Just inform you this is happening, probably best you run an eye over it at some point.

**Comment by HUSSAR-mtrela** [comment]
@aeoncleanse I think  @ChrisKitching should review this  PR when it is complete. So far I have minor improvements and UI thread blocking should be resolved when I complete the last task.

File: lua/system/utils.lua [hunk 1]
```diff
@@ -316,43 +327,76 @@ end
 --- table.reverse {'one','two','three'} => {'three', 'two', 'one'}
 function table.reverse(t)
     if not t then return {} end -- prevents looping over nil table
-    local reversed = {}
+    local r = {}
```
Comment (Sheeo) [inline lua/system/utils.lua]: I don't see why substituting `reversed` for `r` was required here? This is certainly less descriptive.

Not that I generally mind succinct algorithms, but just shortening variable names for the sake of it seems redundant.

File: lua/system/utils.lua [hunk 2]
```diff
@@ -339,11 +339,14 @@ function table.indexize(t)
...
 
 --- Converts a table to a new table with values as keys and values equal to true
+--- it is useful for quickly looking up values in tables instead of looping over it 
+--- table.hash { [1] = 'A',  [2] = 'B',  [3] = 'C' } => 
+--             { [A] = true, [B] = true, [C] = true } 
 function table.hash(t)
     if not t then return {} end -- prevents looping over nil table
     local lookup = {}
     for k, v in t do
-        lookup[tostring(v)] = true
```
Comment (Sheeo) [inline lua/system/utils.lua]: This also means that you'll get peculiar errors when you're trying to hashify a table containing values of type `number | string`. Consider this:

```lua
local t = {[1] = 'A', [2] = 1, [3] = 3}
table.hash(t)
=> { 'A' => true, [3] = true }
```

Your 30% performance figure isn't backed by enough data (How many iterations did you run?).

File: lua/system/utils.lua (hunk 1 above)
Comment (HUSSAR-mtrela) [inline lua/system/utils.lua]: I found it confusing or difficult to understand when names of variables have similar names of functions. Also, these change makes names of variables consistent with rests of variables in this file.

File: lua/system/utils.lua (hunk 2 above)
Comment (HUSSAR-mtrela) [inline lua/system/utils.lua]: @Sheeo you brought valid point here and I will see if there is other way of creating a key for hash table that would not require using tostring() function.

I calculated my performance figures with 10 iterations on tables with 15000+ items. Also @Uveso notice that table.hash(t) was noticeably slower when it was using tostring() function.

**Comment by HUSSAR-mtrela** [comment]
Closing this PR because it is obsoleted by PR#1766

**Pull request closed without merging**